### ISSUE TITLE
Logger report `callback_mode` for `gen_statem` changed in OTP27

### ISSUE THREAD

**Issue by sabiwara** [erlang/otp#8605]
**Describe the bug**

Since this [change](https://github.com/erlang/otp/commit/07eff3b4ca2b0c8b4d42cb8a1a77933b81ab1502#diff-34ee198a2316c68cf9385c73ef4664dfbab1931ce3a14c636ae4c5880c5680f4R2530), the `logger:report()` emitted by when `gen_statem` crashes (when using `callback_mode=handle_event_function`)  has a different value for the `callback_mode` key:
- it used to be the `handle_event_function` atom in OTP 26-
- it is now the module's `fun Mod:handle_event/4`

Not sure if this is a bug or intentional, feel free to close if it's the latter.
Just reporting in case it wasn't.

**To Reproduce**

(sorry in Elixir, taken from https://github.com/elixir-lang/elixir/pull/13684/files)

```elixir
  defmodule MyGenStatemFormatStatus do
    @behaviour :gen_statem

    @impl true
    def callback_mode, do: :handle_event_function

    @impl true
    def init(state) do
      {:ok, :no_state, state}
    end

    @impl true
    def handle_event({:call, _}, :error, :no_state, _data) do
      raise "oops"
    end
  end
```

**Expected behavior**
Perhaps it should be consistent with OTP26- and return `handle_event_function`?

**Affected versions**

OTP 27.0

**Comment by RaimoNiskanen** [comment]
I agree that it is a bug introduced when caching of callback functions was implemented.  The logger map `callback_mode` value should be reverted to `handle_event_function`.

Thank you for reporting this!

**Comment by RaimoNiskanen** [comment]
Completed by #8648 to be released in OTP 27.1
